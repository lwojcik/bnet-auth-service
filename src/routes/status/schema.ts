const response = {
  200: {
    type: "object",
    properties: {
      status: { type: "number" },
      message: { type: "string" },
      timestamp: { type: "string" },
    },
  },
  400: {
    type: "object",
    properties: {
      status: { type: "number" },
    },
  },
  500: {
    type: "object",
    properties: {
      status: { type: "number" },
    },
  },
};

const schema = {
  response,
};

export default schema;
