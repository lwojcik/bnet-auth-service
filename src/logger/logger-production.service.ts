import { Injectable } from '@nestjs/common';
import { InjectPinoLogger, PinoLogger } from 'nestjs-pino';

@Injectable()
export class LoggerProductionService {
  constructor(@InjectPinoLogger() private readonly pinoLogger: PinoLogger) {}

  public setLoggedClass() {
    return false;
  }

  public setLoggedMethod() {
    return false;
  }

  public debug() {
    return false;
  }

  public error(obj: unknown, msg?: string, ...args) {
    return this.pinoLogger.error(obj, msg, args);
  }
}
