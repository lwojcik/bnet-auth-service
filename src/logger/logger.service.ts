/* istanbul ignore file */
import { Injectable, Scope } from '@nestjs/common';
import { InjectPinoLogger, PinoLogger } from 'nestjs-pino';
import { APP } from '../common/constants';
import { Environment } from '../common/types';

@Injectable({
  scope:
    // eslint-disable-next-line no-nested-ternary
    process.env[APP.environment] !== Environment.production
      ? process.env[APP.environment] === Environment.test
        ? Scope.DEFAULT
        : Scope.TRANSIENT
      : Scope.DEFAULT,
})
export class LoggerService {
  private loggedClass: string;

  private loggedMethod: string;

  constructor(@InjectPinoLogger() private readonly pinoLogger: PinoLogger) {}

  private getPrefix() {
    return `${this.loggedClass}${
      this.loggedMethod ? `.${this.loggedMethod}` : ''
    }`;
  }

  private isPrefixSet() {
    return this.loggedClass || this.loggedMethod;
  }

  private formatArgs(args: unknown) {
    if (typeof args === 'string' && args.length > 0) {
      return `'${args}`;
    }
    if (typeof args === 'string' && args.length === 0) {
      return '';
    }
    if (typeof args === 'object' && Object.values(args).length > 0) {
      return JSON.stringify(args);
    }
    return '';
  }

  public setLoggedClass(className: string) {
    this.loggedClass = className;
  }

  public setLoggedMethod(methodName: string, args?: unknown) {
    this.loggedMethod = `${methodName}(${this.formatArgs(args)})`;
  }

  public error(obj: unknown) {
    return this.pinoLogger.error(obj);
  }

  public debug(msg?: unknown, ...args) {
    if (this.isPrefixSet()) {
      this.pinoLogger.debug(
        `${this.getPrefix()}${msg ? `: ${msg}` : ''}`,
        args
      );
    }
  }

  public warn(msg: unknown) {
    return this.pinoLogger.warn(msg);
  }

  public info(msg: string) {
    return this.pinoLogger.info(msg);
  }
}
