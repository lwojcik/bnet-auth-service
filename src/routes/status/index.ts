import { FastifyPluginCallback } from "fastify";
import fp from "fastify-plugin";
import schema from "./schema";

// eslint-disable-next-line no-empty-pattern
const route: FastifyPluginCallback<Record<string, never>> = (
  server,
  // eslint-disable-next-line no-empty-pattern
  {},
  next
) => {
  // eslint-disable-next-line no-empty-pattern
  server.get("/status", { schema }, ({}, reply) => {
    reply.code(200).send({
      status: 200,
      message: "ok",
      timestamp: new Date(),
    });
  });
  next();
};

export default fp(route);
