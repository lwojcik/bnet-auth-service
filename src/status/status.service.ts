import { Injectable } from '@nestjs/common';
import { Duration } from 'luxon';
import { StatusResponse } from './dto/status-response.dto';

@Injectable()
export class StatusService {
  getStatus(): StatusResponse {
    const uptimeSeconds = Math.floor(process.uptime());
    const uptimeMinutes = Math.floor(uptimeSeconds / 60);
    const uptimeHours = Math.floor(uptimeMinutes / 60);
    const uptimeDays = Math.floor(uptimeHours / 24);

    return {
      status: 'ok',
      uptime: Duration.fromObject({
        days: uptimeDays,
        hours: uptimeHours,
        minutes: uptimeMinutes,
      }).toHuman({
        unitDisplay: 'narrow',
      }),
      timestamp: new Date().toISOString(),
    };
  }
}
