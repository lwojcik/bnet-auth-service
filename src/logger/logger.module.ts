import { Module } from '@nestjs/common';
import { LoggerModule as PinoLoggerModule } from 'nestjs-pino';
import { v4 } from 'uuid';
import { LoggerService } from './logger.service';

@Module({
  imports: [
    PinoLoggerModule.forRoot({
      pinoHttp: {
        // prettier-ignore
        genReqId: // istanbul ignore next
          () => v4().toString(),
        level:
          // eslint-disable-next-line no-nested-ternary
          process.env.NODE_ENV === 'production'
            ? // istanbul ignore next
              'error'
            : process.env.NODE_ENV === 'development'
            ? // istanbul ignore next
              'debug'
            : 'silent',
        transport:
          process.env.NODE_ENV !== 'production'
            ? {
                target: 'pino-pretty',
                options: {
                  messageFormat: '[{req.url}] {res.status} {msg}',
                  colorize: true,
                  translateTime: 'SYS:dd/mm/yyyy, HH:MM:ss',
                  ignore: 'req,res,pid,context,responseTime',
                  singleLine: true,
                },
              }
            : // istanbul ignore next
              undefined,
      },
    }),
  ],
  providers: [LoggerService],
  exports: [LoggerService],
})
export class LoggerModule {}
