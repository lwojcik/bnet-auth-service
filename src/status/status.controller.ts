import { Controller, Get } from '@nestjs/common';
import { StatusService } from './status.service';

@Controller('status')
export class StatusController {
  constructor(private readonly statusService: StatusService) {}

  @Get()
  getStatus() {
    return this.statusService.getStatus();
  }
}
