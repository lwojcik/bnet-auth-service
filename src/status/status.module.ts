import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { authConfig } from '../config';
import { LoggerModule } from '../logger/logger.module';
import { StatusController } from './status.controller';
import { StatusService } from './status.service';

@Module({
  imports: [LoggerModule, ConfigModule.forFeature(authConfig)],
  controllers: [StatusController],
  providers: [StatusService],
})
export class StatusModule {}
