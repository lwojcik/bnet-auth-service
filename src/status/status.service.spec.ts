import { Test, TestingModule } from '@nestjs/testing';
import { LoggerService } from '../logger/logger.service';
import { StatusService } from './status.service';

describe('StatusService', () => {
  let service: StatusService;
  let logger: LoggerService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        {
          provide: LoggerService,
          useValue: {
            debug: jest.fn(),
            log: jest.fn(),
            error: jest.fn(),
            setLoggedClass: jest.fn(),
            setLoggedMethod: jest.fn(),
          },
        },
        StatusService,
      ],
    }).compile();

    service = module.get<StatusService>(StatusService);
    logger = module.get<LoggerService>(LoggerService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should return status response', () => {
    jest.spyOn(process, 'uptime').mockImplementation(() => 128);
    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

    expect(service.getStatus()).toMatchSnapshot();
  });

  it('getStatus() should generate logs', () => {
    service.getStatus();
    expect(logger.setLoggedClass).toHaveBeenCalledWith(StatusService.name);
    expect(logger.setLoggedMethod).toHaveBeenCalledWith(
      StatusService.prototype.getStatus.name
    );
    expect(logger.debug).toHaveBeenCalledTimes(1);
  });
});
