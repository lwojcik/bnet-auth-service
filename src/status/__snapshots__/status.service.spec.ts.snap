// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusService should return status response 1`] = `
{
  "status": "ok",
  "timestamp": "2020-01-01T00:00:00.000Z",
  "uptime": "0d, 0h, 2m",
}
`;
